akka-projection-slick/src/main/scala/akka/projection/slick/internal/SlickOffsetStore.scala

```diff
@@ -23,39 +24,73 @@ import slick.jdbc.JdbcProfile
     val db: P#Backend#Database,
     val profile: P,
     clock: Clock) {
+  import OffsetSerialization.MultipleOffsets
+  import OffsetSerialization.SingleOffset
   import OffsetSerialization.fromStorageRepresentation
   import OffsetSerialization.toStorageRepresentation
   import profile.api._
 
-  def this(db: P#Backend#Database, profile: P) = this(db, profile, Clock.systemUTC())
+  def this(db: P#Backend#Database, profile: P) =
+    this(db, profile, Clock.systemUTC())
 
   def readOffset[Offset](projectionId: ProjectionId)(implicit ec: ExecutionContext): Future[Option[Offset]] = {
     val action =
-      offsetTable.filter(_.projectionId === projectionId.id).result.headOption.map { maybeRow =>
-        maybeRow.map(row => fromStorageRepresentation[Offset](row.offsetStr, row.manifest))
+      offsetTable.filter(_.projectionName === projectionId.name).result.map { maybeRow =>
+        maybeRow.map(row => SingleOffset(projectionId, row.manifest, row.offsetStr, row.mergeable))
       }
 
-    db.run(action)
+    val results = db.run(action)
+
+    results.map {
+      case Nil => None
+      case reps if reps.forall(_.mergeable) =>
+        Some(fromStorageRepresentation[MergeableOffsets.Offset[_], Offset](MultipleOffsets(reps)).asInstanceOf[Offset])
+      case reps =>
+        reps.find(_.id == projectionId) match {
+          case Some(rep) => Some(fromStorageRepresentation[Offset, Offset](rep))
+          case _         => None
+        }
+    }
   }
 
+  private def newRow[Offset](rep: SingleOffset, now: Instant): DBIO[_] =
+    offsetTable.insertOrUpdate(OffsetRow(rep.id.name, rep.id.key, rep.offsetStr, rep.manifest, rep.mergeable, now))
+
   def saveOffset[Offset](projectionId: ProjectionId, offset: Offset)(
-      implicit ec: ExecutionContext): slick.dbio.DBIO[Done] = {
-    val (offsetStr, manifest) = toStorageRepresentation(offset)
-    val now = Instant.now(clock)
-    offsetTable.insertOrUpdate(OffsetRow(projectionId.id, offsetStr, manifest, now)).map(_ => Done)
+      implicit ec: ExecutionContext): slick.dbio.DBIO[_] = {
+    val now: Instant = Instant.now(clock)
+    toStorageRepresentation(projectionId, offset) match {
+      case offset: SingleOffset  => newRow(offset, now)
+      case MultipleOffsets(reps) =>
+        // TODO: maybe there's a more "slick" way to accumulate DBIOs like this?
```

Comment:
`DBIOAction.sequence` ?
https://scala-slick.org/doc/3.2.0/dbio.html#sequential-execution

Reply:
I was playing around with this last week and ran into some strange implicit discovery errors, but I think this was just a false positive compile error raised by IntelliJ.  I've updated the impl. to use DBIO.sequence.